File: src/advent/2023/day2.clj
(ns advent.2023.day2
  "Advent of Code 2023, day 2: Cube Conundrum"
  (:require [advent.helpers :as h]
            [clojure.string :as str]))

(def puzzle-input (h/slurp-resource "2023/day2.txt" h/slurp-lines))

(defn parse-round [str-round]
  (->> (str/split str-round #"(,)? ")
       (partition 2)
       (reduce (fn [acc [count color]] (assoc acc (keyword color) (h/parse-int count))) {})))

(defn parse-game
  "Transforms string input line into a list of {:color count} maps."
  [str-game]
  (->> (str/split str-game #"[:;] ")
       (rest)
       (map parse-round)))

;;; Part 1 - Counting the games with rounds that contain less than max-vals items.

(def max-vals {:red 12 :green 13 :blue 14})

(defn valid-round? [round]
  (every? (fn [[color count]] (<= count (color max-vals))) round))

(defn valid-game? [rounds]
  (every? valid-round? rounds))

(defn puzzle1 [input]
  (apply + (map-indexed (fn [idx l] (if (valid-game? (parse-game l)) (inc idx) 0)) input)))

;;; Part 2 - Finding the largest amount of cubes needed per game, then multiplying the values.

(defn puzzle2 [input]
  (apply + (map #(->> % parse-game (apply merge-with max) vals (apply *)) input)))

File: src/haskell/2023/Day2.hs
{-|
Module      : Day2
Description : Day 2: Cube Conundrum

<https://adventofcode.com/2023/day/2>
-}

module Day2 where

import Advent

-- Transforms ["3",  "red", "4", "blue"] into [('r', 3), ('b', 4)]
pairs :: [String] -> [(Char, Int)]
pairs (x:y:xs) = (head y, read x) : pairs xs
pairs _ = []

validGame :: [(Char, Int)] -> Bool
validGame [] = True
validGame (('r', n):xs) | n < 13 = validGame xs
validGame (('g', n):xs) | n < 14 = validGame xs
validGame (('b', n):xs) | n < 15 = validGame xs
validGame _ = False

maxColorValue :: (Int, Int, Int) -> (Char, Int) -> (Int, Int, Int)
maxColorValue (r,g,b) ('r', n) | r < n = (n, g, b)
maxColorValue (r,g,b) ('g', n) | g < n = (r, n, b)
maxColorValue (r,g,b) ('b', n) | b < n = (r, g, n)
maxColorValue acc _ = acc

tupleProduct :: (Int,Int, Int) -> Int
tupleProduct (a,b,c) = a * b * c

main = do
    input <- parsedInput (2023, 2) (map (pairs . drop 2 . words) . lines)
    print $ sum $ map snd $ filter (validGame . fst) $ zip input [1..]
    print $ sum $ map (tupleProduct . foldl maxColorValue (0,0,0)) input
